src/sources/kafka.rs

```diff
@@ -370,105 +411,559 @@ impl SourceConfig for KafkaSourceConfig {
 
 async fn kafka_source(
     config: KafkaSourceConfig,
-    consumer: StreamConsumer<CustomContext>,
+    consumer: StreamConsumer<KafkaSourceContext>,
     decoder: Decoder,
-    mut shutdown: ShutdownSignal,
-    mut out: SourceSender,
-    acknowledgements: bool,
+    out: SourceSender,
+    shutdown: ShutdownSignal,
+    eof: bool,
     log_namespace: LogNamespace,
 ) -> Result<(), ()> {
+    let span = info_span!("kafka_source");
     let consumer = Arc::new(consumer);
-    let (finalizer, mut ack_stream) =
-        OrderedFinalizer::<FinalizerEntry>::maybe_new(acknowledgements, Some(shutdown.clone()));
-    let finalizer = finalizer.map(Arc::new);
-    if let Some(finalizer) = &finalizer {
-        consumer
-            .context()
-            .finalizer
-            .set(Arc::clone(finalizer))
-            .expect("Finalizer is only set once");
+
+    consumer
+        .context()
+        .consumer
+        .set(Arc::downgrade(&consumer))
+        .expect("Error setting up consumer context.");
+
+    let (callback_sender, callback_rx) = mpsc::unbounded_channel();
+    consumer
+        .context()
+        .callbacks
+        .set(callback_sender)
+        .expect("Error setting up consumer callback channel.");
+
+    // EOF signal allowing the coordination task to tell the kafka client task when all partitions have reached EOF
+    let (eof_tx, eof_rx) = if eof {
+        let (tx, rx) = oneshot::channel::<()>();
+        (Some(tx), Some(rx))
+    } else {
+        (None, None)
+    };
```

Comment:
FWIW Due to a recent change to the `Option` type, it can now `unzip` tuples:
```suggestion
    let (eof_tx, eof_rx) = eof.then(|| oneshot::channel::<()>()).unzip();
```